app/assets/javascripts/discourse/app/mixins/quote-button.js

```diff
@@ -0,0 +1,296 @@
+import Mixin from "@ember/object/mixin";
+import { propertyEqual } from "discourse/lib/computed";
+import { selectedElement, selectedText } from "discourse/lib/utilities";
+import { INPUT_DELAY } from "discourse-common/config/environment";
+import KeyEnterEscape from "discourse/mixins/key-enter-escape";
+import { action } from "@ember/object";
+import discourseDebounce from "discourse-common/lib/debounce";
+import { schedule } from "@ember/runloop";
+import toMarkdown from "discourse/lib/to-markdown";
+import { regexSafeStr } from "discourse/lib/quote";
+
+export default Mixin.create(KeyEnterEscape, {
+  classNames: ["quote-button"],
+  classNameBindings: ["visible", "_displayFastEditInput:fast-editing"],
+  visible: false,
+  editText: null,
+
+  _isFastEditable: false,
+  _displayFastEditInput: false,
+  _fastEditInitalSelection: null,
+  _fastEditNewSelection: null,
+  _isSavingFastEdit: false,
+  _canEditPost: false,
+
+  _isMouseDown: false,
+  _reselected: false,
+
+  _hideButton() {
+    this.quoteState.clear();
+    this.set("visible", false);
+
+    this.set("_isFastEditable", false);
+    this.set("_displayFastEditInput", false);
+    this.set("_fastEditInitalSelection", null);
+    this.set("_fastEditNewSelection", null);
+  },
+
+  _selectionChanged() {
+    if (this._displayFastEditInput) {
+      return;
+    }
+
+    const quoteState = this.quoteState;
+
+    const selection = window.getSelection();
+    if (selection.isCollapsed) {
+      if (this.visible) {
+        this._hideButton();
+      }
+      return;
+    }
+
+    // TODO (martin) Allow for > 1 element's worth of content for chat,
+    // post only allows one to be selected.
+    let firstRange, requiredDataForQuote;
+    for (let r = 0; r < selection.rangeCount; r++) {
+      const range = selection.getRangeAt(r);
+      const $selectionStart = $(range.startContainer);
+      const $ancestor = $(range.commonAncestorContainer);
+
+      if (this._noCloseQuotableEl($selectionStart)) {
+        return;
+      }
+
+      firstRange = firstRange || range;
+
+      requiredDataForQuote = this._getRequiredQuoteData(
```

Comment:
Feels a little odd conceptually that the mixin calls `this._getRequiredQuoteData` but `_getRequiredQuoteData` is only defined in the `topic-quote-button` component. Same for `_noCloseContentEl` and `_findCooked` below. Maybe we should abstract this into one `getQuoteData` function that each component must have. (And we can have a default implementation in the mixin that posts a console warning if the component is missing its own `getQuoteData`).

Reply:
I see what you mean, it is a little bit clunky but these functions really are specific to the quote button of the area that we are quoting from:

```javascript
// code related to extraction of quote from elements
  //
  // this must be overridden for any other different type
  // of quote buttons that we implement, because everything
  // can have different data needed to form the quote and
  // different CSS classes etc.
  _getRequiredQuoteData($ancestor, requiredData = {}) {
    if (requiredData.postId) {
      return requiredData;
    }
    return { postId: $ancestor.closest(".boxed, .reply").data("post-id") };
  },

  _hasRequiredQuoteData(requiredQuoteData) {
    return requiredQuoteData.postId !== null;
  },

  _noCloseContentEl($ancestor) {
    return $ancestor.closest(".contents").length === 0;
  },

  _noCloseQuotableEl($selectionStart) {
    return $selectionStart.closest(".cooked").length === 0;
  },

  _findCooked($selectedElement) {
    return (
      $selectedElement.find(".cooked")[0] ||
      $selectedElement.closest(".cooked")[0]
    );
  },
```

Perhaps we just include all of these inside the `quote-button` mixin as empty methods then throw not implemented errors e.g.

```javascript
_findCooked($selectedElement) {
  throw new Error("not implemented");
},
```

The chat usage of this will require further customization than the topic one (I think).